kolena/workflow/_thresholded.py

```diff
@@ -0,0 +1,41 @@
+# Copyright 2021-2023 Kolena Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from abc import ABCMeta
+from dataclasses import dataclass
+from dataclasses import fields
+
+from kolena.workflow._datatypes import TypedDataObject
+
+
+class PreventThresholdOverrideMeta(ABCMeta, type):
+    def __new__(cls, name, bases, dct):
+        if "threshold" in dct.get("__annotations__", {}):
+            for base in bases:
+                if base.__name__ == "ThresholdedMetrics":
+                    raise TypeError(f"Subclasses of {base.__name__} cannot override 'threshold'")
+        return super().__new__(cls, name, bases, dct)
+
+
+@dataclass(frozen=True)
+class ThresholdedMetrics(TypedDataObject, metaclass=PreventThresholdOverrideMeta):
```

Comment:
Sorry, last change here -- since this will be public to our users, lets add a docstring about usage.

Reply:
Added

Reply:
Would prefer to have this a bit more detailed / user actionable (example: [nested test case metrics](https://github.com/kolenaIO/kolena/blob/trunk/kolena/workflow/evaluator.py#L76-L104)).

I'd structure this similarly:
```
    A data class representing metrics associated with a specified threshold.

    `List[ThresholdedMetrics]` is intended for usage as a field type for a [`MetricsTestSample`][kolena.workflow.MetricsTestSample] or [`MetricsTestCase`][kolena.workflow.MetricsTestCase]. This list is intended to expected to contain threshold-specific metric values for a discrete set of specific thresholds that are consistent across [`TestSample`][kolena.workflow.TestSample]s within a test run.

    `ThresholdedMetrics` does not allow dictionary objects as field values and ensures
    that the threshold cannot be overridden.
    It should be subclassed with the relevant fields for a given workflow.

    Example usage:

    ```python
    from kolena.workflow import MetricsTestSample
    from kolena.workflow import ThresholdedMetrics
    from kolena.workflow.annotations import BoundingBox

    @dataclass(frozen=True)
    class ClassThresholdedMetrics(ThresholdedMetrics):
        precision: float
        recall: float
        f1: float

    @dataclass(frozen=True)
    class TestSampleMetrics(MetricsTestSample):
        car: List[ClassThresholdedMetrics]
        pedestrian: List[ClassThresholdedMetrics]

    # metrics object construction
    metric = TestSampleMetrics(
        car=[
            ClassThresholdedMetrics(
                threshold=0.3,
                precision=0.5,
                recall=0.8,
                f1=0.615,
            ),
            ClassThresholdedMetrics(
                threshold=0.4,
                precision=0.6,
                recall=0.6,
                f1=0.6,
            ),
            ClassThresholdedMetrics(
                threshold=0.5,
                precision=0.8,
                recall=0.4,
                f1=0.533,
            ),
            ...
        ],
        pedestrian=[
            ClassThresholdedMetrics(
                threshold=0.3,
                precision=0.6,
                recall=0.9,
                f1=0.72,
            ),
            ClassThresholdedMetrics(
                threshold=0.4,
                precision=0.7,
                recall=0.7,
                f1=0.7,
            ),
            ClassThresholdedMetrics(
                threshold=0.5,
                precision=0.8,
                recall=0.6,
                f1=0.686,
            ),
            ...
        ],
    )
    ```

    Raises:
        TypeError: If any field value is a dictionary.
```